lib/src/executable.dart

```diff
@@ -57,130 +56,152 @@ Future<int> runExecutable(Entrypoint entrypoint, String package,
 
   entrypoint.migrateCache();
 
-  // Unless the user overrides the verbosity, we want to filter out the
-  // normal pub output that may be shown when recompiling snapshots if we are
-  // not attached to a terminal. This is to not pollute stdout when the output
-  // of `pub run` is piped somewhere.
-  return await log.warningsOnlyUnlessTerminal(() async {
-    // Uncached packages are run from source.
-    if (snapshotPath != null) {
-      // Since we don't access the package graph, this doesn't happen
-      // automatically.
-      entrypoint.assertUpToDate();
-
-      var result = await _runOrCompileSnapshot(snapshotPath, args,
-          packagesFile: packagesFile,
-          enableAsserts: enableAsserts,
-          recompile: recompile);
-      if (result != null) return result;
+  var snapshotPath = entrypoint.snapshotPathOfExecutable(executable);
+
+  // Don't compile snapshots for mutable packages, since their code may
+  // change later on.
+  //
+  // Also we don't snapshot if we have non-default arguments to the VM, as
+  // these would be inconsistent if another set of settings are given in a
+  // later invocation.
+  var useSnapshot =
+      !entrypoint.packageGraph.isPackageMutable(package) && vmArgs.isEmpty;
+
+  var executablePath = entrypoint.resolveExecutable(executable);
+  if (!fileExists(executablePath)) {
+    var message = 'Could not find ${log.bold(executable.relativePath)}';
+    if (entrypoint.isGlobal || package != entrypoint.root.name) {
+      message += ' in package ${log.bold(package)}';
     }
+    log.error('$message.');
+    return exit_codes.NO_INPUT;
+  }
 
-    // If the command has a path separator, then it's a path relative to the
-    // root of the package. Otherwise, it's implicitly understood to be in
-    // "bin".
-    if (p.split(executable).length == 1) executable = p.join('bin', executable);
-
-    var executablePath = await _executablePath(entrypoint, package, executable);
+  if (useSnapshot) {
+    // Since we don't access the package graph, this doesn't happen
+    // automatically.
+    entrypoint.assertUpToDate();
 
+    if (!fileExists(snapshotPath)) {
+      await recompile(executable);
+    }
+    if (!fileExists(snapshotPath)) {
+      // TODO XXX compilation failed, run from source??
+    }
```

Comment:
I think `dart.snapshot` will throw `ApplicationException` if the compilation fails.. SP there is no need to handle that here.